### PR TITLE
Implement SQL Unnest

### PR DESCRIPTION
**Pull request by mbudiu-vmw**

### CODE REVIEW

File: SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/optimize/Simplify.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import org.dbsp.sqlCompiler.ir.expression.DBSPBinaryExpression;
+import org.dbsp.sqlCompiler.ir.expression.DBSPExpression;
+import org.dbsp.sqlCompiler.ir.expression.DBSPIfExpression;
+import org.dbsp.sqlCompiler.ir.expression.literal.DBSPBoolLiteral;
+import org.dbsp.sqlCompiler.ir.expression.literal.DBSPIsNullExpression;
+import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBool;
+
+import java.util.Objects;
+
+public class Simplify extends InnerExpressionRewriteVisitor {
```
Comment (lalithsuresh) [inline SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/optimize/Simplify.java:36]: This class could use some comments.

File: SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java [hunk 2]
```diff
@@ -208,13 +197,99 @@ public DBSPAggregate createAggregate(RelNode node,
...
+            index++;
+        }
+        throw new RuntimeException("Type " + type + " has no field named " + field);
+    }
+
     public void visitCorrelate(LogicalCorrelate correlate) {
-        throw new Unimplemented(correlate);
+        // We decorrelate queries using Calcite's optimizer.
+        // So we assume that the only correlated queries we receive
+        // are unnest-type queries.  We assume that unnest queries
```
Comment (lalithsuresh) [inline SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java:217]: So we cannot have correlated subqueries?

File: SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/ArrayTests.java [hunk 3]
```diff
@@ -44,20 +58,238 @@ public void testArray() {
...
+
+        this.addRustTestCase(circuit, new InputOutputPair(
+                new DBSPZSetLiteral[0], new DBSPZSetLiteral[]{ result }));
+    }
+
+    //@Test
+    public void testUnnest2() {
+        // TODO: the Calcite type seems to be wrong, it has no nullable columns
+        DBSPCompiler compiler = testCompiler();
+        String query = "CREATE VIEW V AS SELECT * FROM UNNEST(ARRAY [1, 2, 3, 4, 5], ARRAY[3, 2, 1])";
```
Comment (lalithsuresh) [inline SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/ArrayTests.java:183]: All the tests seem to use integer arrays. Can there be TEXT arrays too?

**Review by lalithsuresh — CHANGES_REQUESTED** [review]

File: SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java (hunk 2 above)
Comment (mbudiu-vmw) [reply SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java:217]: Yes, they are decorrelated by Calcite.

File: SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/ArrayTests.java (hunk 3 above)
Comment (mbudiu-vmw) [reply SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/ArrayTests.java:183]: There will be more tests from Postgres. This just contains very simple tests.

**Pull request merged**